From: Andi Kleen <ak@linux.intel.com>
Date: Fri, 4 Dec 2015 03:50:32 -0800
Subject: perf/x86: Use INST_RETIRED.TOTAL_CYCLES_PS for cycles:pp for Skylake
Git-commit: 442f5c74cbeaf54939980397ece59360c0a824e9
Patch-mainline: v4.5-rc1
References: FATE#321269 (for context)
Signed-off-by: Tony Jones <tonyj@suse.de>

    perf/x86: Use INST_RETIRED.TOTAL_CYCLES_PS for cycles:pp for Skylake
    
    I added UOPS_RETIRED.ALL by mistake to the Skylake PEBS event list for
    cycles:pp. But the event is not documented for Skylake, and has some
    issues.
    
    The recommended replacement for cycles:pp is to use
    INST_RETIRED.ANY+pebs as a base, similar to what CPUs before Sandy
    Bridge did. This new event is called INST_RETIRED.TOTAL_CYCLES_PS. The
    event is not really new, but has been already used by perf before
    Sandy Bridge for the original cycles:p
    
    Note the SDM doesn't document that event either, but it's being
    documented in the latest version of the event list on:
    
      https://download.01.org/perfmon/SKL
    
    This patch does:
    
     - Remove UOPS_RETIRED.ALL from the Skylake PEBS event list
    
     - Add INST_RETIRED.ANY to the Skylake PEBS event list, and an table entry to
       allow cmask=16,inv=1 for cycles:pp
    
     - We don't need an extra entry for the base INST_RETIRED event,
       because it is already covered by the catch-all PEBS table entry.
    
     - Switch Skylake to use the Core2 PEBS alias (which is
       INST_RETIRED.TOTAL_CYCLES_PS)
    
    Signed-off-by: Andi Kleen <ak@linux.intel.com>
    Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
    Cc: Arnaldo Carvalho de Melo <acme@redhat.com>
    Cc: Jiri Olsa <jolsa@redhat.com>
    Cc: Linus Torvalds <torvalds@linux-foundation.org>
    Cc: Peter Zijlstra <peterz@infradead.org>
    Cc: Stephane Eranian <eranian@google.com>
    Cc: Thomas Gleixner <tglx@linutronix.de>
    Cc: Vince Weaver <vincent.weaver@maine.edu>
    Cc: hpa@zytor.com
    Link: http://lkml.kernel.org/r/1448929689-13771-1-git-send-email-andi@firstfloor.org
    Signed-off-by: Ingo Molnar <mingo@kernel.org>

---
 arch/x86/kernel/cpu/perf_event_intel.c    |    2 +-
 arch/x86/kernel/cpu/perf_event_intel_ds.c |    5 ++---
 2 files changed, 3 insertions(+), 4 deletions(-)

--- a/arch/x86/kernel/cpu/perf_event_intel.c
+++ b/arch/x86/kernel/cpu/perf_event_intel.c
@@ -3604,7 +3604,7 @@ __init int intel_pmu_init(void)
 		x86_pmu.event_constraints = intel_skl_event_constraints;
 		x86_pmu.pebs_constraints = intel_skl_pebs_event_constraints;
 		x86_pmu.extra_regs = intel_skl_extra_regs;
-		x86_pmu.pebs_aliases = intel_pebs_aliases_snb;
+		x86_pmu.pebs_aliases = intel_pebs_aliases_core2;
 		/* all extra regs are per-cpu when HT is on */
 		x86_pmu.flags |= PMU_FL_HAS_RSP_1;
 		x86_pmu.flags |= PMU_FL_NO_HT_SHARING;
--- a/arch/x86/kernel/cpu/perf_event_intel_ds.c
+++ b/arch/x86/kernel/cpu/perf_event_intel_ds.c
@@ -727,9 +727,8 @@ struct event_constraint intel_hsw_pebs_e
 
 struct event_constraint intel_skl_pebs_event_constraints[] = {
 	INTEL_FLAGS_UEVENT_CONSTRAINT(0x1c0, 0x2),	/* INST_RETIRED.PREC_DIST */
-	INTEL_FLAGS_UEVENT_CONSTRAINT_DATALA_NA(0x01c2, 0xf), /* UOPS_RETIRED.ALL */
-	/* UOPS_RETIRED.ALL, inv=1, cmask=16 (cycles:p). */
-	INTEL_FLAGS_EVENT_CONSTRAINT(0x108001c2, 0xf),
+	/* INST_RETIRED.TOTAL_CYCLES_PS (inv=1, cmask=16) (cycles:p). */
+	INTEL_FLAGS_EVENT_CONSTRAINT(0x108000c0, 0x0f),
 	INTEL_PLD_CONSTRAINT(0x1cd, 0xf),		      /* MEM_TRANS_RETIRED.* */
 	INTEL_FLAGS_UEVENT_CONSTRAINT_DATALA_LD(0x11d0, 0xf), /* MEM_INST_RETIRED.STLB_MISS_LOADS */
 	INTEL_FLAGS_UEVENT_CONSTRAINT_DATALA_ST(0x12d0, 0xf), /* MEM_INST_RETIRED.STLB_MISS_STORES */
